aten/src/ATen/native/sparse/SparseMatMul.cpp

```diff
@@ -0,0 +1,255 @@
+  auto mat2_values = mat2._values().contiguous();
+  LongTensor mat2_row_indices = mat2_indices_.select(0, 0);
+  LongTensor mat2_col_indices = mat2_indices_.select(0, 1);
+
+  Tensor mat2_indptr = coo_to_csr(mat2_row_indices.data_ptr<int64_t>(), K, mat2._nnz()); 
+
+  auto nnz = _csr_matmult_maxnnz(M, N, mat1_indptr.data_ptr<int64_t>(), mat1_col_indices.data_ptr<int64_t>(), 
+      mat2_indptr.data_ptr<int64_t>(), mat2_col_indices.data_ptr<int64_t>());
+
+  auto output_indices = output._indices();
+  auto output_values = output._values();
+
+  Tensor output_indptr = at::empty({M + 1}, kLong);
+
+  output_indices.resize_({2, nnz});
+  output_values.resize_(nnz);
+
+  LongTensor output_row_indices = output_indices.select(0, 0);
+  LongTensor output_col_indices = output_indices.select(0, 1);
+
+  _csr_matmult(M, N, mat1_indptr.data_ptr<int64_t>(), mat1_col_indices.data_ptr<int64_t>(), mat1_values.data_ptr<scalar_t>(), 
+  mat2_indptr.data_ptr<int64_t>(), mat2_col_indices.data_ptr<int64_t>(), mat2_values.data_ptr<scalar_t>(), 
+  output_indptr.data_ptr<int64_t>(), output_col_indices.data_ptr<int64_t>(), output_values.data_ptr<scalar_t>());
+  
+  csr_to_coo(M, output_indptr.data_ptr<int64_t>(), output_row_indices.data_ptr<int64_t>());
+}
+
+} // end anonymous namespace
+
+Tensor fill_with_ones_cpu(const Tensor& input){
+
+  Tensor output = at::ones(input.sizes(), input.options().layout(kStrided));
+  
+  AT_DISPATCH_FLOATING_TYPES(output.scalar_type(), "fill_with_ones", [&] {
+        auto input_indices = input._indices();
+        auto input_values = input._values();
+        auto nnz = input._nnz();
+        auto input_indices_accessor = input_indices.accessor<int64_t, 2>();
+        auto input_values_accessor = input_values.accessor<scalar_t, 1>();
+
+        auto output_values_accessor = output.accessor<scalar_t, 2>();
+
+        at::parallel_for(0, nnz, 0, [&](int64_t start, int64_t end) {
+          for (auto index = start; index < end; index++) {
+            auto x = input_indices_accessor[0][index];
+            auto y = input_indices_accessor[1][index];
+            output_values_accessor[x][y] = input_values_accessor[index];
+          }
+        });
+      });
+  return output;
+}
+
+Tensor sparse_sparse_matmul_cpu(const Tensor& mat1_, const Tensor& mat2_) {
+  TORCH_INTERNAL_ASSERT(mat1_.is_sparse());
+  TORCH_INTERNAL_ASSERT(mat2_.is_sparse());
+  TORCH_CHECK(mat1_.dim() == 2);
+  TORCH_CHECK(mat2_.dim() == 2);
+
+  TORCH_CHECK(
+      mat1_.size(1) == mat2_.size(0), "mat1 and mat2 shapes cannot be multiplied (",
+      mat1_.size(0), "x", mat1_.size(1), " and ", mat2_.size(0), "x", mat2_.size(1), ")");
+
+  TORCH_CHECK(mat1_.scalar_type() == mat2_.scalar_type(),
+           "mat1 dtype ", mat1_.scalar_type(), " does not match mat2 dtype ", mat2_.scalar_type());
+
+  auto output = at::native::empty_like(mat1_);
+  output.sparse_resize_and_clear_({mat1_.size(0), mat2_.size(1)}, mat1_.sparse_dim(), mat1_.dense_dim());
```

Comment:
What will happen if either of these input tensors is a hybrid tensor? Is that case supported?

Can you elaborate on how this resize is computed using the sparse and dense dims?

Reply:
Hybrid tensor are not supported. I will add a check `TORCH_CHECK(mat1_.dense_dim() == 0, ...` for both inputs.

As `dense_dim` is always `0` the `sparse_resize_and_clear_` method will generate a sparse tensor with `sparse_dim==2` and `dense_dim==0` to represent just a sparse matrix.